File: Problem202.cpp
#include<iostream>

using namespace std;

class Solution {
public:
    bool isHappy(int n) {
        int run = getNext(n), walk = n;
        while(run != 1){
            if(run == walk){return false;}
            run = getNext(getNext(run)), walk = getNext(walk);
        }
        return true;
    }
private:
    int getNext(int n){
        int result = 0;
        while(n > 0){
            result += ((n % 10) * (n % 10));
            n /= 10;
        }
        return result;
    }
};

int main(){
	Solution* solution = new Solution();
	cout << solution->isHappy(19) << " [expected: 1]" << endl;
	cout << solution->isHappy(2) << " [expected: 0]" << endl;
	return 0;
}

File: Problem202.py
#solution for LeetCode problem 202
class Solution:
    def isHappy(self, n: int) -> bool:
        def getNext(number):
            result = 0
            while number > 0:
                number, digit = divmod(number, 10)
                result += digit ** 2
            return result
        walk = n
        run = getNext(n)
        while run != 1:
            if walk == run:
                return False
            run = getNext(getNext(run))
            walk = getNext(walk)
        return True

if __name__ == "__main__":
	solution = Solution()
	print(solution.isHappy(19), "[expected: True]") #should return True
	print(solution.isHappy(2), "[expected: False]") #should return False
